File: src/main/java/no/nav/tjenestepensjon/simulering/soap/SamlHeaderCallback.java
package no.nav.tjenestepensjon.simulering.soap;

import java.util.Base64;

import javax.xml.namespace.QName;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerConfigurationException;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;

import org.springframework.util.Assert;
import org.springframework.ws.WebServiceMessage;
import org.springframework.ws.client.core.WebServiceMessageCallback;
import org.springframework.ws.soap.SoapHeaderElement;
import org.springframework.ws.soap.SoapMessage;
import org.springframework.xml.transform.StringSource;

public class SamlHeaderCallback implements WebServiceMessageCallback {

    private final String token;
    private final Transformer transformer;

    public SamlHeaderCallback(String token) {
        this.token = token;
        try {
            this.transformer = TransformerFactory.newInstance().newTransformer();
        } catch (TransformerConfigurationException e) {
            throw new RuntimeException(e.getMessage(), e);
        }
    }

    @Override
    public void doWithMessage(WebServiceMessage message) throws TransformerException {
        Assert.isInstanceOf(SoapMessage.class, message);
        SoapMessage soapMessage = (SoapMessage) message;
        SoapHeaderElement header =
                soapMessage.getSoapHeader().addHeaderElement(new QName("http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd", "Security"));
        transformer.transform(new StringSource(new String(Base64.getDecoder().decode(token))), header.getResult());
    }
}

File: src/main/kotlin/no/nav/tjenestepensjon/simulering/soap/SamlHeaderCallback.kt
package no.nav.tjenestepensjon.simulering.soap

import no.nav.tjenestepensjon.simulering.config.ApplicationProperties.SAML_SECURITY_CONTEXT_URL
import org.springframework.util.Assert
import org.springframework.ws.WebServiceMessage
import org.springframework.ws.client.core.WebServiceMessageCallback
import org.springframework.ws.soap.SoapMessage
import org.springframework.xml.transform.StringSource
import java.util.*
import javax.xml.namespace.QName
import javax.xml.transform.TransformerConfigurationException
import javax.xml.transform.TransformerException
import javax.xml.transform.TransformerFactory

class SamlHeaderCallback(private val token: String?) : WebServiceMessageCallback {
    private val transformer = try {
        TransformerFactory.newInstance().newTransformer()
    } catch (e: TransformerConfigurationException) {
        throw RuntimeException(e.message, e)
    }

    @Throws(TransformerException::class)
    override fun doWithMessage(message: WebServiceMessage) {
        Assert.isInstanceOf(SoapMessage::class.java, message)
        val soapMessage = message as SoapMessage
        transformer.transform(
                StringSource(String(Base64.getDecoder().decode(token))),
                soapMessage.soapHeader.addHeaderElement(QName(SAML_SECURITY_CONTEXT_URL, "Security")).result
        )
    }
}